### PR TITLE
[TVMScript] ExprDoc

### PR DESCRIPTION
**Pull request by yelite**
This PR addes:

- All ExprDoc subclasses
- Their Python bindings
- Support of ExprDoc in PythonDocPrinter
- Unit tests for ExprDoc in PythonDocPrinter

Tracking issue: https://github.com/apache/tvm/issues/11912

cc @junrushao1994 @gbonik

### CODE REVIEW

File: include/tvm/script/printer/doc.h [hunk 1]
```diff
@@ -63,13 +63,43 @@ class Doc : public ObjectRef {
...
+  ExprDoc Call(Array<ExprDoc, void> args) const;
+
+  /*!
+   * \brief Create a doc representing attribute access on the current ExprDoc
+   * \param args The positional arguments of the function call.
+   * \param kwargs_keys Keys of keywords arguments of the function call.
+   * \param kwargs_values Values of keywords arguments of the function call.
+   */
+  ExprDoc Call(Array<ExprDoc, void> args,        //
+               Array<String, void> kwargs_keys,  //
```
Comment (junrushao1994) [inline include/tvm/script/printer/doc.h:100]: nit

```suggestion
               Array<String> kwargs_keys,  //
```

**Comment by yelite** [comment]
I removed the Assert op as we are going to make assert a StmtDoc per dicussion with @junrushao1994. I also added an if-then-else special op

File: include/tvm/script/printer/doc.h [hunk 2]
```diff
@@ -158,6 +188,454 @@ class LiteralDoc : public ExprDoc {
...
+    kEq,        // ==
+    kNotEq,     // !=
+    kGt,        // >
+    kGtE,       // >=
+    kBinaryEnd,
+
+    // Special
+    kSpecialStart,
+    kIfThenElse,  // <operands[1]> if <operands[0]> else <operands[2]>
+    kSpecialEnd
```
Comment (junrushao1994) [inline include/tvm/script/printer/doc.h:401]: Let's explicitly specify their values:

```suggestion
    // Unary operators
    kUnaryStart = 0,
    kUSub = 1,    // -x
    kInvert = 2,  // ~x
    kUnaryEnd = 3,

    // Binary operators
    kBinaryStart,
    kAdd,       // +
    kSub,       // -
    kMult,      // *
    kDiv,       // /
    kFloorDiv,  // // in Python
    kMod,       // % in Python
    kPow,       // ** in Python
    kLShift,    // <<
    kRShift,    // >>
    kBitAnd,    // &
    kBitOr,     // |
    kBitXor,    // ^
    kLt,        // <
    kLtE,       // <=
    kEq,        // ==
    kNotEq,     // !=
    kGt,        // >
    kGtE,       // >=
    kBinaryEnd,

    // Special
    kSpecialStart,
    kIfThenElse,  // <operands[1]> if <operands[0]> else <operands[2]>
    kSpecialEnd
```

File: include/tvm/script/printer/doc.h [hunk 3]
```diff
@@ -158,6 +188,454 @@ class LiteralDoc : public ExprDoc {
...
+ * \sa OperationDoc
+ */
+class OperationDocNode : public ExprDocNode {
+ public:
+  enum class Kind : int32_t {
+    // Unary operators
+    kUnaryStart,
+    kUSub,    // -x
+    kInvert,  // ~x
+    kUnaryEnd,
```
Comment (junrushao1994) [inline include/tvm/script/printer/doc.h:374]: Let's explicitly specify their values:

```suggestion
    // Unary operators
    kUnaryStart = 0,
    kUSub = 1,    // -x
    kInvert = 2,  // ~x
    kUnaryEnd = 3,
```

File: include/tvm/script/printer/doc.h [hunk 4]
```diff
@@ -158,6 +188,454 @@ class LiteralDoc : public ExprDoc {
...
+ * This doc can only appear in IndexDoc::indices.
+ *
+ * \sa AttrAccessDoc
+ */
+class SliceDocNode : public DocNode {
+ public:
+  /*! \brief The start of slice */
+  Optional<ExprDoc> start;
+  /*! \brief The exclusive end of slice */
+  Optional<ExprDoc> stop;
```
Comment (junrushao1994) [inline include/tvm/script/printer/doc.h:611]: We would probably need to support `step` given it's used in TVMScript: https://github.com/apache/tvm/pull/11308.

CC @cyx-6

File: python/tvm/script/printer/doc.py [hunk 5]
```diff
@@ -29,12 +34,63 @@ class Doc(Object):
 class ExprDoc(Object):
     """Base class of all expression Docs"""
 
+    def attr_access(self, attr: str) -> "AttrAccessDoc":
```
Comment (junrushao1994) [inline python/tvm/script/printer/doc.py:37]: Let's be consist with C++ APIs

```suggestion
    def attr(self, attr: str) -> "AttrAccessDoc":
```

File: python/tvm/script/printer/doc.py [hunk 6]
```diff
@@ -29,12 +34,63 @@ class Doc(Object):
...
+        attr : str
+            The attribute name to access
+
+        Returns
+        -------
+        doc : AttrAccessDoc
+        """
+        return _ffi_api.ExprDocAttr(self, attr)  # type: ignore
+
+    def index_access(self, indices: List[Union["ExprDoc", "SliceDoc"]]) -> "IndexDoc":
```
Comment (junrushao1994) [inline python/tvm/script/printer/doc.py:52]: ditto

```suggestion
    def index(self, indices: List[Union["ExprDoc", "SliceDoc"]]) -> "IndexDoc":
```

File: python/tvm/script/printer/doc.py [hunk 7]
```diff
@@ -29,12 +34,63 @@ class Doc(Object):
...
+        indices : List[Union["ExprDoc", "SliceDoc"]]
+            The indices to access
+
+        Returns
+        -------
+        doc : IndexDoc
+        """
+        return _ffi_api.ExprDocIndex(self, indices)  # type: ignore
+
+    def call_with(self, *args: Tuple["ExprDoc"], **kwargs: Dict[str, "ExprDoc"]) -> "CallDoc":
```
Comment (junrushao1994) [inline python/tvm/script/printer/doc.py:67]: ditto

```suggestion
    def call(self, *args: Tuple["ExprDoc"], **kwargs: Dict[str, "ExprDoc"]) -> "CallDoc":
```

File: python/tvm/script/printer/doc.py [hunk 8]
```diff
@@ -47,3 +103,171 @@ def __init__(self, value):
...
+    def __init__(self, value: ExprDoc, attr: str):
+        self.__init_handle_by_constructor__(_ffi_api.AttrAccessDoc, value, attr)  # type: ignore
+
+
+@tvm._ffi.register_object("script.printer.IndexDoc")
+class IndexDoc(ExprDoc):
+    """Doc that represents index access on an expression"""
+
+    value: ExprDoc
+    indices: tvm.ir.container.Array  # actual type:  List[Union[ExprDoc, "SliceDoc"]]
```
Comment (junrushao1994) [inline python/tvm/script/printer/doc.py:134]: Simply using the actual type

```suggestion
    indices: List[Union[ExprDoc, "SliceDoc"]]
```

File: python/tvm/script/printer/doc.py [hunk 9]
```diff
@@ -47,3 +103,171 @@ def __init__(self, value):
...
+
+
+@tvm._ffi.register_object("script.printer.CallDoc")
+class CallDoc(ExprDoc):
+    """Doc that represents function call"""
+
+    callee: ExprDoc
+    args: tvm.ir.container.Array  # actual type: List[ExprDoc]
+    kwargs_keys: tvm.ir.container.Array  # actual type: List[str]
+    kwargs_values: tvm.ir.container.Array  # actual type: List[ExprDoc]
```
Comment (junrushao1994) [inline python/tvm/script/printer/doc.py:147]: ditto

File: python/tvm/script/printer/doc.py [hunk 10]
```diff
@@ -47,3 +103,171 @@ def __init__(self, value):
...
+    LtE = auto()
+    Eq = auto()
+    NotEq = auto()
+    Gt = auto()
+    GtE = auto()
+    _BinaryEnd = auto()
+
+    _SpecialStart = auto()
+    IfThenElse = auto()
+    _SpecialEnd = auto()
```
Comment (junrushao1994) [inline python/tvm/script/printer/doc.py:196]: Let's be explicitly consistent with numbering on C++ side. Avoid using `auto`

File: python/tvm/script/printer/doc.py [hunk 11]
```diff
@@ -47,3 +103,171 @@ def __init__(self, value):
...
+class OperationDoc(ExprDoc):
+    """
+    Doc that represents operation
+
+    It can be unary, binary and other special operators (for example, the
+    if-then-else expression).
+    """
+
+    kind: OperationKind
+    operands: tvm.ir.container.Array  # actual type: List[ExprDoc]
```
Comment (junrushao1994) [inline python/tvm/script/printer/doc.py:211]: ditto

File: src/script/printer/python_doc_printer.cc [hunk 12]
```diff
@@ -57,6 +82,179 @@ void PythonDocPrinter::PrintTypedDoc(const LiteralDoc& doc) {
...
+    output_ << "[()]";
+  } else {
+    output_ << "[";
+    PrintJoinedDocs(doc->indices, ", ");
+    output_ << "]";
+  }
+}
+
+const std::string OperatorToString(OperationDocNode::Kind operation_kind) {
+  static const std::vector<std::string> OP_STR_TABLE = []() {
```
Comment (junrushao1994) [inline src/script/printer/python_doc_printer.cc:104]: nit: no need to case this table given it's just a local variable (it's just personal preference)

```suggestion
  static const std::vector<std::string> op_kind2str = []() {
```

File: src/script/printer/python_doc_printer.cc [hunk 13]
```diff
@@ -57,6 +82,179 @@ void PythonDocPrinter::PrintTypedDoc(const LiteralDoc& doc) {
...
+    PrintDoc(doc->operands[0]);
+  } else if (doc->kind < OpKind::kBinaryEnd) {
+    // Binary Operator
+    ICHECK_EQ(doc->operands.size(), 2);
+    PrintDoc(doc->operands[0]);
+    output_ << " " << OperatorToString(doc->kind) << " ";
+    PrintDoc(doc->operands[1]);
+  } else if (doc->kind == OpKind::kIfThenElse) {
+    CHECK_EQ(doc->operands.size(), 3)
+        << "ValueError: IfThenElse requires 3 operands, but got " << doc->operands.size();
```
Comment (junrushao1994) [inline src/script/printer/python_doc_printer.cc:165]: You may use ICHECK to be consistent with previous ones

File: src/script/printer/python_doc_printer.cc [hunk 14]
```diff
@@ -57,6 +82,179 @@ void PythonDocPrinter::PrintTypedDoc(const LiteralDoc& doc) {
...
+  }();
+
+  auto op_index = static_cast<int>(operation_kind);
+  ICHECK_LT(op_index, OP_STR_TABLE.size());
+  const std::string str = OP_STR_TABLE[op_index];
+  if (str.empty()) {
+    LOG(FATAL) << "OperationDocNode::Kind " << static_cast<int>(operation_kind)
+               << " cannot be converted to operator token in Python directly.";
+    throw;
+  }
```
Comment (junrushao1994) [inline src/script/printer/python_doc_printer.cc:146]: you may use ICHECK directly

```suggestion
  ICHECK(!str.empty()) << "OperationDocNode::Kind " << static_cast<int>(operation_kind)
               << " cannot be converted to operator token in Python directly.";
  }
```

File: src/script/printer/python_doc_printer.cc [hunk 15]
```diff
@@ -57,6 +82,179 @@ void PythonDocPrinter::PrintTypedDoc(const LiteralDoc& doc) {
...
+  for (const ExprDoc& arg : doc->args) {
+    if (is_first) {
+      is_first = false;
+    } else {
+      output_ << ", ";
+    }
+    PrintDoc(arg);
+  }
+
+  // Print keyword args
```
Comment (junrushao1994) [inline src/script/printer/python_doc_printer.cc:193]: add an `ICHECK_EQ` here for `|kwargs_keys|` and `|kwargs_values|`

File: src/script/printer/python_doc_printer.cc [hunk 16]
```diff
@@ -57,6 +82,179 @@ void PythonDocPrinter::PrintTypedDoc(const LiteralDoc& doc) {
...
+    output_ << ",";
+  } else {
+    PrintJoinedDocs(doc->elements, ", ");
+  }
+  output_ << ")";
+}
+
+void PythonDocPrinter::PrintTypedDoc(const DictDoc& doc) {
+  output_ << "{";
+  size_t idx = 0;
```
Comment (junrushao1994) [inline src/script/printer/python_doc_printer.cc:235]: ditto. add an `ICHECK_EQ` to ensure `|values| == |keys|`

File: src/script/printer/python_doc_printer.cc [hunk 17]
```diff
@@ -57,6 +82,179 @@ void PythonDocPrinter::PrintTypedDoc(const LiteralDoc& doc) {
...
+    }
+    PrintDoc(key);
+    output_ << ": ";
+    PrintDoc(doc->values[idx]);
+    idx++;
+  }
+  output_ << "}";
+}
+
+void PythonDocPrinter::PrintTypedDoc(const SliceDoc& doc) {
```
Comment (junrushao1994) [inline src/script/printer/python_doc_printer.cc:248]: let's support strided slice: `[ : : ]`

**Review by junrushao1994 — APPROVED** [review]
Thanks @yelite for the PR, especially carefully crafted tests! I left a few nitpicks but overall it looks pretty good!

File: include/tvm/script/printer/doc.h (hunk 4 above)
Comment (cyx-6) [reply include/tvm/script/printer/doc.h:611]: Got it!

File: python/tvm/script/printer/doc.py (hunk 5 above)
Comment (yelite) [reply python/tvm/script/printer/doc.py:37]: This is a tricky one. If strictly following C++ APIs and Python name convention, we will have
```python
doc.attr  # helper method to create AttrAccessDoc
doc.attr  # AttrAccessDoc's `attr` attribute, that's a name conflict

doc.index  # helper method to create IndexDoc
doc.indices  # attribute of IndexDoc, looks similar to the method above
```

An option is to update the C++ API to `AttrAccess` and `IndexAccess`, which makes the code slightly more verbose
```cpp
TIR(p)->Attr("cast")
// becomes
TIR(p)->AttrAccess("cast")
```

What do you think?

Comment (yelite) [reply python/tvm/script/printer/doc.py]: Sounds good, updated according to your recommendation

**Pull request merged**